### PR TITLE
Make fast parser handles type annotations + type comments better

### PR DESCRIPTION
**Pull request by afrieder**
This PR has two parts:

1. Disallows two return annotations (ie addresses #2733)
2. Enforces that type annotations and the type comment are consistent.

I assume 1 is not controversial, but 2 might be.

The point of it is to prevent:
```python
def f(x: str):
  # type: (int) -> int
  return x
```

which is currently perfectly allowed. 

Thus, this code checks, for every argument with a type annotation, that the type in the type comment is identical.
The way the code works, `# type: (...) -> int` will _always_ pass. 
The reason this might controversial is that it does string comparison, so it's not always right. In fact, it's very simple to get a false positive:

```python
def f(x: Union[int, str]):
  # type: (Union[str, int]) -> int
  return 5
```

will now complain about an inconsistent type signature. 

Now, I feel this is fine, since if you're going to go through the trouble of using both a type annotation and a non `...` type comment, then you might as well make them the same. The only reason this might be an actual problem is for something like:

```python
T = Union[int, str]
def f(x: Union[int, str]):
  # type: (T) -> int
  return 5
```

Which fails. Because we only keep a single source of truth for what the arg types are, we cannot postpone this check until we have actual types without a large and unnecessary (imo) refactor. 

So in summary, I think it's perfectly fine to insist that if users are going to use both type annotations and type comments in a function that they match identically.
But if the Powers That Be disagree, I'll happily remove the second part from this PR.

### CODE REVIEW

**Comment by afrieder** [comment]
I should also add that because of how the typed AST handles type comments, 

```python
def f(x: int):
  return x
```
and
```python
def f(x  # type: int
   ):
  return x
```
are indistinguishable. 

This is... annoying, because in the old parser 
```python
def f(x: int):
  # type: (...) -> int
  return x
```
is allowed, while
```python
def f(x  # type: int
   ):
  # type: (...) -> 5
  return x
```
is not.

Since we cannot replicate this behavior, I chose to just enforce that the two methods of annotating convey the same type information.

**Comment by ddfisher** [comment]
Thanks for the PR!  To my understanding, there's no reason anyone should be mixing type comments and type annotations, so it would be better to disallow mixing them entirely rather than to check for consistency.  I agree that it would be better if typed_ast distinguished between per-argument type comments and type annotations, but I don't think that's hugely important.  Because of that limitation, let's just disallow type annotations for all non-ellipsis-parameter type comments.

**Comment by gvanrossum** [comment]
Indeed. PEP 484 says "It is not allowed for an argument or return value to
have both a type annotation and a type comment"

File: mypy/fastparse.py [hunk 1]
```diff
@@ -283,9 +284,15 @@ def do_func_def(self, n: Union[ast35.FunctionDef, ast35.AsyncFunctionDef],
...
                 if (len(func_type_ast.argtypes) == 1 and
                         isinstance(func_type_ast.argtypes[0], ast35.Ellipsis)):
+                    if n.returns:
+                        # PEP 484 disallows both type annotations and type comments
+                        self.fail(messages.DUPLICATE_TYPE_SIGNATURES, n.lineno, n.col_offset)
                     arg_types = [a.type_annotation if a.type_annotation is not None else AnyType()
                                  for a in args]
                 else:
+                    # PEP 484 disallows both type annotations and type comments
+                    if n.returns or any(a.type_annotation for a in args):
```
Comment (ddfisher) [inline mypy/fastparse.py]: It's slightly better to explicitly check that `a.type_annotation is not None` than just to check the truthiness here.

File: mypy/fastparse2.py [hunk 2]
```diff
@@ -284,6 +285,9 @@ def visit_FunctionDef(self, n: ast27.FunctionDef) -> Statement:
                     arg_types = [a.type_annotation if a.type_annotation is not None else AnyType()
                                 for a in args]
                 else:
+                    # PEP 484 disallows both type annotations and type comments
+                    if any(a.type_annotation for a in args):
```
Comment (ddfisher) [inline mypy/fastparse2.py]: (and here also)

File: mypy/fastparse.py (hunk 1 above)
Comment (afrieder) [inline mypy/fastparse.py]: Fixed (in both places).
_Completely unrelated_, sorry about clogging AppVeyor.

**Comment by afrieder** [comment]
Cleaned up the commits, should be good to go? It's all much cleaner now.

**Comment by ddfisher** [comment]
Looks good, thanks!  I'll merge once the tests finish running.

FYI we do squash merges on PRs, so (aside from having things be understandable during review) you don't need to worry about a nice commit history.

**Comment by afrieder** [comment]
Ack.
It was purely for review legibility.
